stellargraph/data/explorer.py

```diff
@@ -107,6 +81,114 @@ def run(self, **kwargs):
+
+    def _check_parameter_values(self, nodes=None, n=0, length=0, seed=None):
+        """
+        Checks that the parameter values are valid or raises ValueError exceptions with a message indicating the
+        parameter (the first one encountered in the checks) with invalid value.
+
+        Args:
+            nodes: <list> A list of root node ids such that from each node a uniform random walk of up to length l
+            will be generated.
+            n: <int> Number of walks per node id.
+            length: <int> Maximum length of walk measured as the number of edges followed from root node.
+            seed: <int> Random number generator seed
+        Returns:
+            The random state as determined by the seed.
+        """
+        self._check_nodes(nodes)
+        self._check_repetitions(n)
+        self._check_length(length)
+        return self._check_seed(seed)
+
+    def _check_nodes(self, nodes):
+        if nodes is None:
+            self._raise_error("A list of root node IDs was not provided.")
+        if not is_real_iterable(nodes):
+            self._raise_error("Nodes parameter should be an iterable of node IDs.")
+        if (
+            len(nodes) == 0
+        ):  # this is not an error but maybe a warning should be printed to inform the caller
+            print(
+                "({}) WARNING: No root node IDs given. An empty list will be returned as a result.".format(
+                    type(self).__name__
+                )
+            )
+
+    def _check_repetitions(self, n):
+        if type(n) != int:
+            self._raise_error(
+                "The number of walks per root node, n, should be integer type."
+            )
+        if n <= 0:
+            self._raise_error(
+                "The number of walks per root node, n, should be a positive integer."
+            )
+
+    def _check_length(self, length):
+        if type(length) != int:
+            self._raise_error("The walk length, length, should be integer type.")
+        if length <= 0:
+            # Technically, length 0 should be okay, but by consensus is invalid.
+            self._raise_error("The walk length, length, should be a positive integer.")
+
+    def _check_seed(self, seed):
+        if seed is None:
+            # Restore the random state
+            return self._random_state
+        if type(seed) != int:
+            self._raise_error(
+                "The random number generator seed value, seed, should be integer type or None."
+            )
+        if seed < 0:
+            self._raise_error(
+                "The random number generator seed value, seed, should be non-negative integer or None."
+            )
+        # seed the random number generator
+        return random.Random(seed)
+
+    # For neighbourhood sampling
+    def _check_sizes(self, n_size):
+        err_msg = "The neighbourhood size must be a list of non-negative integers."
+        if n_size is None or type(n_size) != list:
```

Comment:
Does this need to be a list? Can it be a subclass of a list? I think that should be accepted, e.g. https://docs.python.org/3/library/functions.html#type

> The `isinstance()` built-in function is recommended for testing the type of an object, because it takes subclasses into account.

```suggestion
        if n_size is None or not isinstance(n_size, list):
```

(This also applies to other `type(...)` calls in this code.)

---

In addition, the `None` check is redundant (`type(None)` is `<class 'NoneType'>` (`!= list`), and it isn't a subclass of `list` either):

```suggestion
        if not isinstance(n_size, list):
```

Finally, can this be a tuple? I would expect so? Do we need to check the type here, vs. just checking its length and then iterating?

Reply:
Another good catch (but actually not my code or my design - the basic deign assumes lists everywhere; erk).

Reply:
What's wrong with expecting lists? What else would you pass there?

Reply:
The general principle would be an iterator (if we didn't care about length). If you think in terms of java, Python's list is a concrete class, which breaks the D in SOLID (dependency inversion, i.e. depend on abstractions rather than concrete classes).

Reply:
> What's wrong with expecting lists? What else would you pass there?

Concretely, I could easily imagine passing a tuple here: `(10, 20)` instead of `[10, 20]`.